source/Halibut.Tests/Transport/Streams/TimeoutStreamWrapperFixture.cs

```diff
@@ -0,0 +1,369 @@
+            {
+                await performListenerWrite("Test");
+
+                var buffer = new byte[19];
+                var readBytes = await sut.ReadAsync(buffer, 0, 19, CancellationToken);
+                if (readBytes > 0)
+                {
+                    readData = Encoding.UTF8.GetString(buffer, 0, readBytes);
+                }
+
+                Assert.AreEqual("Test", readData);
+            }
+        }
+
+        [Test]
+        public async Task ReadAsyncShouldCancel()
+        {
+            using (var readTokenSource = new CancellationTokenSource(TimeSpan.FromSeconds(5)))
+            {
+                var (disposables, sut, _) = await BuildTcpClientAndTcpListener(
+                    CancellationToken,
+                    onListenerRead: async _ => await Task.Delay(TimeSpan.FromDays(1), CancellationToken));
+
+                using (disposables)
+                {
+                    // Ensure the timeouts are not used
+                    sut.WriteTimeout = (int)TimeSpan.FromSeconds(120).TotalMilliseconds;
+                    sut.ReadTimeout = (int)TimeSpan.FromSeconds(120).TotalMilliseconds;
+                    
+                    var stopWatch = Stopwatch.StartNew();
+                    Exception? actualException = null;
+
+                    // Brute force attempt to get the Write to be slow
+                    while (!CancellationToken.IsCancellationRequested)
+                    {
+                        try
+                        {
+                            await sut.ReadAsync(new byte[19], 0, 19, readTokenSource.Token);
+                        }
+                        catch (Exception e)
+                        {
+                            actualException = e;
+                            break;
+                        }
+                    }
+                    stopWatch.Stop();
+
+                    actualException.Should().NotBeNull().And.BeOfType<OperationCanceledException>();
+                    actualException!.Message.Should().Be("The ReadAsync operation was cancelled.");
+
+                    stopWatch.Elapsed.Should().BeLessThan(TimeSpan.FromSeconds(10));
+                }
+            }
+        }
+
+        [Test]
+        public async Task ReadAsyncShouldTimeout()
+        {
+            var (disposables, sut, _) = await BuildTcpClientAndTcpListener(
+                CancellationToken, 
+                onListenerRead: async _ => await Task.Delay(TimeSpan.FromDays(1), CancellationToken));
+
+            using (disposables)
+            {
+                // Ensure the correct timeout is used
+                sut.WriteTimeout = (int)TimeSpan.FromSeconds(120).TotalMilliseconds;
+                sut.ReadTimeout = (int)TimeSpan.FromSeconds(5).TotalMilliseconds;
+
+                var stopWatch = Stopwatch.StartNew();
+                Exception? actualException = null;
+
+                // Brute force attempt to get the Write to be slow
```

Comment:
Nit: Seeing this pattern a few times, I wonder if it might be nicer to have a method like `RunTillException` that takes a lambda and CancellationToken, and returns the exception. Or even takes a generic argument of the expected exception type. 

Then we could use it for reads and writes.

Just a thought.

Reply:
We have a Try.CatchingError helper that I will adapt to async and use here to remove boilerplate